core/src/apresolve.rs

```diff
@@ -1,132 +1,124 @@
-        Some((host, port))
-    });
-
-    let ap = if ap_port.is_some() {
-        aps.find(|(_, p)| *p == port)
-    } else {
-        aps.next()
-    };
-
-    ap.unwrap_or_else(|| (String::from(fallback), port))
+// These addresses probably do some geo-location based traffic management or at least DNS-based
+// load balancing. They are known to fail when the normal resolvers are up, so that's why they
+// should only be used as fallback.
+impl Default for ApResolveData {
+    fn default() -> Self {
+        Self {
+            accesspoint: vec![String::from("ap.spotify.com:443")],
+            dealer: vec![String::from("dealer.spotify.com:443")],
+            spclient: vec![String::from("spclient.wg.spotify.com:443")],
+        }
+    }
 }
 
-async fn try_apresolve(proxy: Option<&Url>) -> Result<ApResolveData, Box<dyn Error>> {
-    let req = Request::builder()
-        .method("GET")
-        .uri(APRESOLVE_ENDPOINT)
-        .body(Body::empty())
-        .unwrap();
-
-    let response = if let Some(url) = proxy {
-        // Panic safety: all URLs are valid URIs
-        let uri = url.to_string().parse().unwrap();
-        let proxy = Proxy::new(Intercept::All, uri);
-        let connector = HttpConnector::new();
-        let proxy_connector = ProxyConnector::from_proxy_unsecured(connector, proxy);
-        Client::builder()
-            .build(proxy_connector)
-            .request(req)
-            .await?
-    } else {
-        Client::new().request(req).await?
-    };
-
-    let body = hyper::body::to_bytes(response.into_body()).await?;
-    let data: ApResolveData = serde_json::from_slice(body.as_ref())?;
-
-    Ok(data)
+component! {
+    ApResolver : ApResolverInner {
+        data: AccessPoints = AccessPoints::default(),
+    }
 }
 
-pub async fn apresolve(proxy: Option<&Url>, ap_port: Option<u16>) -> AccessPoints {
-    let data = try_apresolve(proxy).await.unwrap_or_else(|e| {
-        warn!("Failed to resolve access points: {}, using fallbacks.", e);
-        ApResolveData::default()
-    });
-
-    let accesspoint = select_ap(data.accesspoint, AP_FALLBACK, ap_port);
-    let dealer = select_ap(data.dealer, DEALER_FALLBACK, ap_port);
-    let spclient = select_ap(data.spclient, SPCLIENT_FALLBACK, ap_port);
+impl ApResolver {
+    fn split_aps(data: Vec<String>) -> Vec<SocketAddress> {
+        data.into_iter()
+            .filter_map(|ap| {
+                let mut split = ap.rsplitn(2, ':');
+                let port = split
+                    .next()
+                    .expect("rsplitn should not return empty iterator");
+                let host = split.next()?.to_owned();
+                let port: u16 = port.parse().ok()?;
+                Some((host, port))
+            })
+            .collect()
+    }
 
-    AccessPoints {
-        accesspoint,
-        dealer,
-        spclient,
+    fn find_ap(&self, data: &[SocketAddress]) -> usize {
+        match self.session().config().proxy {
```

Comment:
Don't check for `proxy` but `ap_port`. This way, users can configure to use a certain port even without a proxy. This is useful for example when port 4070 is firewalled and you want to force traffic over 443.

Reply:
Fixed in 3a7843d.